node/network/src/network_handler.rs

```diff
@@ -3,13 +3,41 @@ use protocol::ProtocolHandler;
 use primitives::traits::GenericResult;
 use primitives::types::SignedTransaction;
 use std::sync::Arc;
+use parking_lot::RwLock;
+
+use futures::sync::mpsc;
+use futures::{Future, Sink};
 
 pub struct NetworkHandler {
-    pub client: Arc<Client>,
+    pub client: Arc<RwLock<Client>>,
 }
 
 impl ProtocolHandler for NetworkHandler {
     fn handle_transaction(&self, t: SignedTransaction) -> GenericResult {
-        self.client.handle_signed_transaction(t)
+        self.client.write().handle_signed_transaction(t)
+    }
+}
+
+/// Reports transactions received from a network into the given channel.
+pub struct ChannelNetworkHandler {
+    transactions_sender: mpsc::Sender<SignedTransaction>,
+}
+
+impl ChannelNetworkHandler {
+    pub fn new(transactions_sender: mpsc::Sender<SignedTransaction>) -> Self {
+        Self { transactions_sender }
+    }
+}
+
+impl ProtocolHandler for ChannelNetworkHandler {
+    fn handle_transaction(&self, transaction: SignedTransaction) -> Result<(), &'static str> {
+        let copied_tx = self.transactions_sender.clone();
+        tokio::spawn(
```

Comment:
is it really cheaper to spawn a send? isn't this just putting it on the channel?

Reply:
Send takes some small but noticable time, so we don't want to block on it (I did some experiments in the past).

Things that use ProtocolHandler do not expect the future to be returned. Our code lacks futures, so we cannot return it here, someone should eventually poll that future somewhere else